File: docs/palindromes-in-peg/scaffold_gs_matcher.py
"""Full finite GS matcher with real input queues and frozen prefix heads.

Standalone protocol: load binary letters (draining the bounded broadcast
after each), ! freezes the reversed prefix, then append text letters and
service local instructions with dots. This is a compiler/interfacing test;
the final PAL controller must schedule these internal operations itself.
"""
from gs_match_heads import compile_matcher, MATCH_HEADS, MATCH_TESTS
from gs_head_liveness import analyze_readers
from scaffold_circuit import (Circuit, Value, Ref, NEW, PREVIOUS, TRUE, FALSE,
                              conjunction, disjunction, neg, choose)
from scaffold_live_distances import LiveDistances
from scaffold_stream_heads import StreamBank, PatternTextHead


class GSMatcher:
  def __init__(self, circuit, prefix="matcher"):
    self.circuit, self.prefix = circuit, prefix
    self.program = compile_matcher()
    self.readers = analyze_readers(self.program)
    self.distances = LiveDistances(circuit, self.program, MATCH_HEADS, prefix + ".distance",
                                  observe_positions=False, availability_distance=False)
    names = tuple(f"{prefix}.r{i}" for i in range(self.readers.registers))
    self.streams = StreamBank(circuit, (prefix + ".begin", prefix + ".end",
                                      *(name + suffix for name in names for suffix in (".pattern", ".text"))))
    self.begin = self.streams.heads[prefix + ".begin"]
    self.end = self.streams.heads[prefix + ".end"]
    self.registers = [PatternTextHead(self.streams, name) for name in names]
    self.heads = {head: self.registers[color] for head, color in self.readers.colors.items()}
    self.pc_key, self.phase_key = prefix + ".pc", prefix + ".phase"
    self.pc = circuit.get(PREVIOUS, self.pc_key, tuple(range(len(self.program.code))), self.program.start)
    self.phase = circuit.get(PREVIOUS, self.phase_key, ("idle", "broadcast", "maintenance", "run"), "idle")
    self.active_key = prefix + ".active"
    self.active = circuit.get(PREVIOUS, self.active_key, (False, True), False).eq(True)
    self.queue_key, self.remaining_key = prefix + ".queue", prefix + ".remaining"
    qnames = tuple(self.streams.queues.queues)
    self.queue = circuit.get(PREVIOUS, self.queue_key, qnames, qnames[0])
    self.remaining = circuit.get(PREVIOUS, self.remaining_key, (0, 1, 2, 3), 0)
    self.return_key, self.index_key = prefix + ".return", prefix + ".broadcast_index"
    self.return_phase = circuit.get(PREVIOUS, self.return_key, ("broadcast", "run"), "broadcast")
    self.broadcast_queues = (self.begin.queue.name, *(head.text.queue.name for head in self.registers))
    self.broadcast_index = circuit.get(PREVIOUS, self.index_key, tuple(range(len(self.broadcast_queues) + 1)), 0)
    self.cell_key, self.output_key = prefix + ".input_cell", prefix + ".matched"
    self.cell = circuit.get_ref(PREVIOUS, self.cell_key)
    self.output = circuit.get(PREVIOUS, self.output_key, (False, True), False).eq(True)

  def tick(self, symbols=None):
    c = self.circuit
    char = c.input() if symbols is None else symbols
    arrival = disjunction(char.eq("a"), char.eq("b"))
    start = char.eq("!")
    dot = char.eq(".")
    c.require(disjunction(self.phase.eq("idle"), self.phase.eq("run")), disjunction(arrival, start))
    c.require(self.end.focus.present(), start)
    phase = self.phase
    run = conjunction(dot, phase.eq("run"))
    broadcast = conjunction(dot, phase.eq("broadcast"))
    maintenance = conjunction(dot, phase.eq("maintenance"))
    events = [conjunction(run, self.pc.eq(i)) for i in range(len(self.program.code))]
    next_pc, copies, moves, requests = {}, {}, {}, {}
    matched = FALSE

    def edge(target, guard):
      next_pc[target] = disjunction(next_pc.get(target, FALSE), guard)

    def request(queue, guard):
      requests[queue] = disjunction(requests.get(queue, FALSE), guard)

    for state, (enabled, (event, targets)) in enumerate(zip(events, self.program.code)):
      op, *args = event
      if op in MATCH_TESTS:
        if op == "available":
          decision = self.heads[args[0]].available()
        elif op == "symbols":
          a, b = (self.heads[name].read(enabled) for name in args)
          c.require(conjunction(neg(a.eq(None)), neg(b.eq(None))), enabled)
          decision = a.equal(b)
        else:
          decision = (self.distances.equal if op == "equal" else self.distances.less)(*args)
        edge(targets[0], conjunction(enabled, neg(decision)))
        edge(targets[1], conjunction(enabled, decision))
      elif op == "halt":
        c.require(FALSE, enabled)
      else:
        edge(targets[0], enabled)
        if op == "copy":
          target, source = args
          if target in self.readers.after[state]:
            pair = self.readers.colors[target], self.readers.colors[source]
            if pair[0] != pair[1]:
              copies[pair] = disjunction(copies.get(pair, FALSE), enabled)
        elif op == "move":
          head, direction = args[0][0]
          if head in self.readers.after[state]:
            key = self.readers.colors[head], direction
            moves[key] = disjunction(moves.get(key, FALSE), enabled)
        elif op == "assert_equal":
          c.require(self.distances.equal(*args), enabled)
        elif op == "match":
          # A positive result is for this arrival exactly. A queued older
          # match must not silently become the next prefix's answer.
          c.require(neg(self.heads["B"].available()), enabled)
          matched = disjunction(matched, enabled)
        else:
          raise ValueError("unsupported GS matcher instruction")
    self.distances.execute(events)
    for (register, direction), enabled in moves.items():
      queue, needed = self.registers[register].move(direction, enabled)
      request(queue, needed)
    for (target, source), enabled in copies.items():
      self.registers[target].copy_from(self.registers[source], enabled)

    # One broadcast queue is updated in this instruction. The explicit
    # maintenance state services its three work units before proceeding.
    count = len(self.broadcast_queues)
    finished = choose(self.active, self.broadcast_index.eq(count), self.broadcast_index.eq(1))
    push = conjunction(broadcast, neg(finished))
    for index, name in enumerate(self.broadcast_queues):
      enabled = conjunction(push, self.broadcast_index.eq(index))
      self.streams.queues.queues[name].push(self.cell, enabled)
      request(name, enabled)
    for name, queue in self.streams.queues.queues.items():
      queue.work_unit(conjunction(maintenance, self.queue.eq(name)))
    self.end.follow_arrival(NEW, arrival)
    self.distances.load_one(arrival)
    self.cell = Ref.select(arrival, NEW, self.cell)

    self.heads["Origin"].start(self.end, start)
    self.heads["Tail"].start(self.begin, start)
    self.distances.initialize({("Origin", "Tail"): (self.distances.length, -1)}, start)
    self.pc = Value.select(run, Value(next_pc), self.pc)
    self.pc = Value.select(start, Value.constant(self.program.start), self.pc)
    self.active = disjunction(self.active, start)
    self.output = disjunction(conjunction(self.output, neg(disjunction(arrival, start))), matched)
    self.broadcast_index = Value.select(push, self.broadcast_index.map(lambda n: min(n + 1, count)), self.broadcast_index)
    self.broadcast_index = Value.select(arrival, Value.constant(0), self.broadcast_index)

    finished_work = conjunction(maintenance, self.remaining.eq(1))
    self.remaining = Value.select(maintenance, self.remaining.map(lambda n: max(0, n - 1)), self.remaining)
    self.phase = Value.select(finished_work, self.return_phase, self.phase)
    self.phase = Value.select(conjunction(broadcast, finished),
                             Value.select(self.active, Value.constant("run"), Value.constant("idle")), self.phase)
    needed = disjunction(*requests.values())
    self.queue = Value.select(needed, Value(requests), self.queue)
    self.remaining = Value.select(needed, Value.constant(3), self.remaining)
    self.return_phase = Value.select(needed, Value.select(push, Value.constant("broadcast"), Value.constant("run")), self.return_phase)
    self.phase = Value.select(needed, Value.constant("maintenance"), self.phase)
    self.phase = Value.select(arrival, Value.constant("broadcast"), self.phase)
    self.phase = Value.select(start, Value.constant("run"), self.phase)

  def finalize(self):
    self.distances.finalize()
    self.streams.finalize()
    for head in self.registers:
      head.finalize()
    for key, value in ((self.pc_key, self.pc), (self.phase_key, self.phase),
                       (self.queue_key, self.queue), (self.remaining_key, self.remaining),
                       (self.return_key, self.return_phase), (self.index_key, self.broadcast_index)):
      self.circuit.put(key, value)
    for key, value in ((self.active_key, self.active), (self.output_key, self.output)):
      self.circuit.put(key, Value.select(value, Value.constant(True), Value.constant(False)))
    self.circuit.put_ref(self.cell_key, self.cell)


def fixture():
  circuit = Circuit("ab!.")
  circuit.put("input", Value.select(circuit.input().eq("b"), Value.constant("b"), Value.constant("a")), tuple("ab"), "a")
  worker = GSMatcher(circuit)
  worker.tick()
  worker.finalize()
  return circuit, worker, circuit.machine(worker.output)

File: scala/pal/src/main/scala/pal/ScaffoldGsMatcher.scala
package pal

import scala.collection.mutable
import Expr.FALSE
import Ref.{NEW, PREVIOUS}
import ScaffoldCircuit.{conjunction, disjunction, neg, choose}

/** Full finite GS matcher with real input queues and frozen prefix heads.
  * The fixture loads a prefix, freezes it with !, and services instructions
  * with dots. It is a compiler/interface test, not the final PAL grammar.
  * Port of `scaffold_gs_matcher.py`; Python finalize is commit.
  */
final class GSMatcher(val circuit: Circuit, val prefix: String = "matcher") {
  val program: Program = GsMatchHeads.compileMatcher()
  val readers: ReaderLiveness = GsHeadLiveness.analyzeReaders(program)
  val distances = new LiveDistances(circuit, program, GsMatchHeads.MATCH_HEADS, prefix + ".distance",
    observePositions = false, availabilityDistance = false)
  private val names = Vector.tabulate(readers.registers)(i => s"$prefix.r$i")
  val streams = new StreamBank(circuit, Vector(prefix + ".begin", prefix + ".end") ++
    names.flatMap(name => Vector(name + ".pattern", name + ".text")))
  val begin: StreamHead = streams.heads(prefix + ".begin")
  val end: StreamHead = streams.heads(prefix + ".end")
  val registers: Vector[PatternTextHead] = names.map(new PatternTextHead(streams, _))
  val heads: Map[String, PatternTextHead] = readers.colors.map { case (head, color) => head -> registers(color) }
  val pcKey: String = prefix + ".pc"
  val phaseKey: String = prefix + ".phase"
  var pc: Value[Int] = circuit.get(PREVIOUS, pcKey, program.code.indices.toVector, program.start)
  var phase: Value[String] = circuit.get(PREVIOUS, phaseKey, Vector("idle", "broadcast", "maintenance", "run"), "idle")
  val activeKey: String = prefix + ".active"
  var active: Expr = circuit.get(PREVIOUS, activeKey, Vector(false, true), false).eqTo(true)
  val queueKey: String = prefix + ".queue"
  val remainingKey: String = prefix + ".remaining"
  private val queueNames = streams.queues.queues.keys.toVector
  var queue: Value[String] = circuit.get(PREVIOUS, queueKey, queueNames, queueNames.head)
  var remaining: Value[Int] = circuit.get(PREVIOUS, remainingKey, Vector(0, 1, 2, 3), 0)
  val returnKey: String = prefix + ".return"
  val indexKey: String = prefix + ".broadcast_index"
  var returnPhase: Value[String] = circuit.get(PREVIOUS, returnKey, Vector("broadcast", "run"), "broadcast")
  val broadcastQueues: Vector[String] = Vector(begin.queue.name) ++ registers.map(_.text.queue.name)
  var broadcastIndex: Value[Int] = circuit.get(PREVIOUS, indexKey, (0 to broadcastQueues.size).toVector, 0)
  val cellKey: String = prefix + ".input_cell"
  val outputKey: String = prefix + ".matched"
  var cell: Ref = circuit.getRef(PREVIOUS, cellKey)
  var output: Expr = circuit.get(PREVIOUS, outputKey, Vector(false, true), false).eqTo(true)

  def tick(symbols: Option[Value[Any]] = None): Unit = {
    val char = symbols.getOrElse(circuit.input())
    val arrival = disjunction(char.eqTo('a'), char.eqTo('b'))
    val start = char.eqTo('!')
    val dot = char.eqTo('.')
    circuit.require(disjunction(phase.eqTo("idle"), phase.eqTo("run")), disjunction(arrival, start))
    circuit.require(end.focus.present(), start)
    val oldPhase = phase
    val run = conjunction(dot, oldPhase.eqTo("run"))
    val broadcast = conjunction(dot, oldPhase.eqTo("broadcast"))
    val maintenance = conjunction(dot, oldPhase.eqTo("maintenance"))
    val events = program.code.indices.map(i => conjunction(run, pc.eqTo(i))).toVector
    val nextPc = mutable.LinkedHashMap.empty[Int, Expr]
    val copies = mutable.LinkedHashMap.empty[(Int, Int), Expr]
    val moves = mutable.LinkedHashMap.empty[(Int, Int), Expr]
    val requests = mutable.LinkedHashMap.empty[String, Expr]
    var matched = FALSE
    def edge(target: Int, guard: Expr): Unit = { nextPc(target) = disjunction(nextPc.getOrElse(target, FALSE), guard) }
    def request(name: String, guard: Expr): Unit = { requests(name) = disjunction(requests.getOrElse(name, FALSE), guard) }
    for (((enabled, Row(event, targets)), state) <- events.zip(program.code).zipWithIndex) {
      def branch(decision: Expr): Unit = {
        edge(targets(0), conjunction(enabled, neg(decision)))
        edge(targets(1), conjunction(enabled, decision))
      }
      event match {
        case Event.Available(head) => branch(heads(head).available())
        case Event.Symbols(left, right) =>
          val a = heads(left).read(enabled)
          val b = heads(right).read(enabled)
          circuit.require(conjunction(neg(a.eqTo(None)), neg(b.eqTo(None))), enabled)
          branch(a.equal(b))
        case Event.Equal(left, right) => branch(distances.equal(left, right))
        case Event.Less(left, right) => branch(distances.less(left, right))
        case Event.Halt => circuit.require(FALSE, enabled)
        case other =>
          edge(targets(0), enabled)
          other match {
            case Event.Copy(target, source) =>
              if (readers.after(state)(target)) {
                val pair = (readers.colors(target), readers.colors(source))
                if (pair._1 != pair._2) { copies(pair) = disjunction(copies.getOrElse(pair, FALSE), enabled) }
              }
            case Event.Move(batch) =>
              val Event.Movement(head, direction) = batch.head
              if (readers.after(state)(head)) {
                val key = (readers.colors(head), direction)
                moves(key) = disjunction(moves.getOrElse(key, FALSE), enabled)
              }
            case Event.AssertEqual(left, right) => circuit.require(distances.equal(left, right), enabled)
            case Event.Match(_) =>
              circuit.require(neg(heads("B").available()), enabled)
              matched = disjunction(matched, enabled)
            case _ => throw new IllegalArgumentException("unsupported GS matcher instruction")
          }
      }
    }
    distances.execute(events)
    for (((register, direction), enabled) <- moves) {
      val (name, needed) = registers(register).move(direction, enabled)
      request(name, needed)
    }
    for (((target, source), enabled) <- copies) { registers(target).copyFrom(registers(source), enabled) }
    val count = broadcastQueues.size
    val finished = choose(active, broadcastIndex.eqTo(count), broadcastIndex.eqTo(1))
    val push = conjunction(broadcast, neg(finished))
    for ((name, index) <- broadcastQueues.zipWithIndex) {
      val enabled = conjunction(push, broadcastIndex.eqTo(index))
      streams.queues.queues(name).push(cell, enabled)
      request(name, enabled)
    }
    for ((name, target) <- streams.queues.queues) { target.workUnit(conjunction(maintenance, queue.eqTo(name))) }
    end.followArrival(NEW, arrival)
    distances.loadOne(arrival)
    cell = Ref.select(arrival, NEW, cell)
    heads("Origin").start(end, start)
    heads("Tail").start(begin, start)
    distances.initialize(Map(("Origin", "Tail") -> Some((distances.length, -1))), start)
    pc = Value.select(run, Value(nextPc), pc)
    pc = Value.select(start, Value.constant(program.start), pc)
    active = disjunction(active, start)
    output = disjunction(conjunction(output, neg(disjunction(arrival, start))), matched)
    broadcastIndex = Value.select(push, broadcastIndex.map(n => math.min(n + 1, count)), broadcastIndex)
    broadcastIndex = Value.select(arrival, Value.constant(0), broadcastIndex)
    val finishedWork = conjunction(maintenance, remaining.eqTo(1))
    remaining = Value.select(maintenance, remaining.map(n => math.max(0, n - 1)), remaining)
    phase = Value.select(finishedWork, returnPhase, phase)
    phase = Value.select(conjunction(broadcast, finished),
      Value.select(active, Value.constant("run"), Value.constant("idle")), phase)
    val needed = disjunction(requests.values.toSeq*)
    queue = Value.select(needed, Value(requests), queue)
    remaining = Value.select(needed, Value.constant(3), remaining)
    returnPhase = Value.select(needed, Value.select(push, Value.constant("broadcast"), Value.constant("run")), returnPhase)
    phase = Value.select(needed, Value.constant("maintenance"), phase)
    phase = Value.select(arrival, Value.constant("broadcast"), phase)
    phase = Value.select(start, Value.constant("run"), phase)
  }

  def commit(): Unit = {
    distances.commit()
    streams.commit()
    registers.foreach(_.commit())
    for ((key, value) <- Vector[(String, Value[Any])](pcKey -> pc, phaseKey -> phase,
      queueKey -> queue, remainingKey -> remaining, returnKey -> returnPhase, indexKey -> broadcastIndex)) {
      circuit.put(key, value)
    }
    for ((key, value) <- Vector(activeKey -> active, outputKey -> output)) {
      circuit.put(key, Value.select(value, Value.constant(true), Value.constant(false)))
    }
    circuit.putRef(cellKey, cell)
  }
}

object ScaffoldGsMatcher {
  def fixture(): (Circuit, GSMatcher, Scaffold) = {
    val circuit = new Circuit("ab!.")
    circuit.put("input", Value.select(circuit.input().eqTo('b'), Value.constant('b'), Value.constant('a')), Vector('a', 'b'), 'a')
    val worker = new GSMatcher(circuit)
    worker.tick()
    worker.commit()
    (circuit, worker, circuit.machine(worker.output))
  }
}
